grid-client/deployer/deployer.go

```diff
@@ -154,29 +154,58 @@ func (d *Deployer) deploy(
 			}
 			log.Debug().Uint32("Number of public ips", publicIPCount)
 
+			contractRused := false
 			contractID, err := d.substrateConn.CreateNodeContract(d.identity, node, dl.Metadata, hashHex, publicIPCount, newDeploymentSolutionProvider[node])
-			log.Debug().Uint64("CreateNodeContract returned id", contractID)
 			if err != nil {
-				return currentDeployments, errors.Wrapf(err, "failed to create contract on node %d", node)
+				if !strings.Contains(err.Error(), "ContractIsNotUnique") {
+					return currentDeployments, errors.Wrapf(err, "failed to create contract on node %d", node)
+				}
+				contractID, err = d.substrateConn.GetContractWithHash(d.identity, node, []byte(hashHex))
+				if err != nil {
+					return currentDeployments, errors.Wrapf(err, "failed to find existing contract on node %d", node)
+				}
+				contract, err := d.substrateConn.GetContract(contractID)
+				if err != nil {
+					return currentDeployments, errors.Wrapf(err, "failed to get existing contract on node %d", node)
+				}
+				if contract.State.IsDeleted {
+					return currentDeployments, errors.Errorf("contract %d is not active", contractID)
+				}
+				log.Info().
+					Uint32("node", node).
+					Uint64("contractID", contractID).
+					Msg("reusing existing contract")
+				contractRused = true
 			}
-
+			log.Debug().Uint64("returned contract ID", contractID)
 			dl.ContractID = contractID
-			err = client.DeploymentDeploy(ctx, dl)
+
+			// Update deployment with contract ID and send to node
+			err = nodeClient.DeploymentDeploy(ctx, dl)
 			if err != nil {
-				rerr := d.substrateConn.EnsureContractCanceled(d.identity, contractID)
-				if rerr != nil {
-					return currentDeployments, errors.Wrapf(err, "error cancelling contract: %s; you must cancel it manually (id: %d)", rerr, contractID)
+				// If deployment exists, continue as already deployed
+				if !contractRused || !strings.Contains(err.Error(), "exists") {
+					// Other deployment error: cancel contract
+					rerr := d.substrateConn.EnsureContractCanceled(d.identity, dl.ContractID)
```

Comment:
isn't it dangerous here? we can cancel existing ones (will the error check be enough?)
should we check ownership first?